plugins/kubernetes/app/models/kubernetes/build_job_executor.rb

```diff
@@ -0,0 +1,144 @@
+# Execute a job that can build/push Docker images
+# and write progress to the local job output.
+module Kubernetes
+  class BuildJobExecutor
+    WAIT_FOR_JOB = 5.minutes
+    TICK = 3.seconds
+
+    attr_reader :job
+    delegate :id, to: :job
+
+    def initialize(output, job:)
+      @output = output
+      @job = job
+      @registry = {}
+    end
+
+    def execute!(build, project, tag:, push: false, registry:, tag_as_latest: false)
+      job_log = job_name = job_namespace = ""
+      k8s_job = nil
+
+      unless valid_registry_info?(registry)
+        @output.puts "### Registry server should not be empty. Aborting..."
+        return false, job_log
+      end
+      @registry = registry
+
+      k8s_job = job_config(build, project, tag: tag, push: push, tag_as_latest: tag_as_latest)
+      job_name = k8s_job[:metadata][:name]
+      job_namespace = k8s_job[:metadata][:namespace]
+
+      @output.puts "### Running a remote Docker build job: #{job_name}"
+      success, job_log = create_and_wait_for_job(k8s_job, @output)
+
+      message = (success ? "completed successfully" : "failed or timed out")
+      @output.puts "### Remote build job #{job_name} #{message}"
+
+      return success, job_log
+    ensure
+      # Jobs will still be there regardless of their statuses
+      begin
+        @output.puts "### Cleaning up the remote build job"
+        extension_client.delete_job(job_name, job_namespace) if k8s_job
+      rescue KubeException
+        @output.puts "### Failed to clean up the remote build job"
+      end
+    end
+
+    private
+
+    def build_cluster
+      Kubernetes::Cluster.first!
```

Comment:
this most likely needs some tweeking ... something like picking a context / cluster via env vars ... but good enough for now ...